### PR TITLE
CORDA-3755: Switched attachments map to a WeakHashMap

### PR DESCRIPTION
**Pull request by adelel1**
Switched the map holding the attachments to a weak hash map. 

The associated strong reference referring to it is the attachment id, which then gets stored in a URL, which is used by the AttachmentsClassLoader.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 1]
```diff
@@ -374,6 +376,9 @@ object AttachmentURLStreamHandlerFactory : URLStreamHandlerFactory {
         return URL(attachmentScheme, "", -1, id, AttachmentURLStreamHandler)
     }
 
+    @VisibleForTesting
+    fun loadedAttachmentsSize(): Int = loadedAttachments.size
+
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt:377]: I guess the question is will the IOException on line 385 happen more often now and is that bad/handled properly?

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 2]
```diff
@@ -120,6 +121,34 @@ class ClassLoadingUtilsTest {
...
+        val attachment = signedAttachment(jarData)
+        var url: URL? = AttachmentURLStreamHandlerFactory.toUrl(attachment)
+
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+        repeat(10) { System.gc() }
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+        url = null
+        var count = 0
+        while (AttachmentURLStreamHandlerFactory.loadedAttachmentsSize() != 0 && count++ < 100) {
+            System.gc()
```
Comment (mnesbit) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: Do we need a test that forces GC during a verify, or something? To ensure that we know references are held thoughout verify and flows that manipulate attachements?

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 3]
```diff
@@ -374,6 +376,9 @@ object AttachmentURLStreamHandlerFactory : URLStreamHandlerFactory {
         return URL(attachmentScheme, "", -1, id, AttachmentURLStreamHandler)
```
Comment (chrisr3) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: You are now _relying_ on `URL` maintaining a strong reference to this _particular_ `id` String object. The fact that it does so is merely a fortunate implementation detail of this `URL` constructor. This is non-obvious and very scary.

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 4]
```diff
@@ -120,6 +121,34 @@ class ClassLoadingUtilsTest {
...
+        }
+        val attachment = signedAttachment(jarData)
+        var url: URL? = AttachmentURLStreamHandlerFactory.toUrl(attachment)
+
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+        repeat(10) { System.gc() }
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+        url = null
+        var count = 0
+        while (AttachmentURLStreamHandlerFactory.loadedAttachmentsSize() != 0 && count++ < 100) {
```
Comment (chrisr3) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: IIRC `System.gc` merely _requests_ that the JVM schedule some garbage collection. Calling it repeatedly like this is just like hammering the button on a pelican crossing. Have you considered using a `ReferenceQueue`? You could put the `String` object inside a _second_ `WeakReference` that has an associated `ReferenceQueue`, clear the strong reference, invoke `System.gc` once and then wait for the object to appear on the queue.

More specifically, you'd probably need to extend `WeakReference<URL>` so that you could add an extra data field to it. `WeakReference.get()` will return `null` once the URL has been garbage collected, but if you remove an instance of your extended `WeakReference` class off the queue then you can still read its extra data field.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 1 above)
Comment (adelel1) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt:377]: It shouldn't happen any more often, actually it shouldn't happen at all.

The existing code doesn't really completely handle a new 'attachments' protocol. i.e. I cannot create a string "attachment::/123454322" pass this to a URL and then stream data from the inputStream. I can only create a URL by passing the attachment I want to stream from. So the life of the attachment matches the life of the URL, this is then passed to URL class loader.
So basically the implementation assumes its only used by the AttachmentsClassLoader.

**Review by chrisr3 — COMMENTED** [review]
I'm sure that I had a branch somewhere that planned to add `/$id` as the URL path, because I had a reason for wanting the path elements to start with `/`. Something to do with matching code-sources to grants in a security policy file (I think). I don't seem to have that branch any more, but that change would break this PR wide open.

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt (hunk 2 above)
Comment (adelel1) [reply core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: Added a test for this for the verify. For flows this change is orthogonal. The change only affects code where the attachments are in the classpath.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 3 above)
Comment (adelel1) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Ok, I have also now attached the lifetime of the attachments in the map to the lifetime of the attachments classloader using them (which should also match the URL lifetime).

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt (hunk 4 above)
Comment (adelel1) [reply core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: Ok - now checking the reference queue after a gc.

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 5]
```diff
@@ -356,24 +366,26 @@ object AttachmentsClassLoaderBuilder {
...
             AttachmentURLStreamHandler
         } else null
     }
 
-    fun toUrl(attachment: Attachment): URL {
-        val id = attachment.id.toString()
-        loadedAttachments[id] = attachment
-        return URL(attachmentScheme, "", -1, id, AttachmentURLStreamHandler)
+    fun toUrl(attachmentWithKey: AttachmentWithKey): URL {
+        loadedAttachments[attachmentWithKey.key] = attachmentWithKey.attachment
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: There is significant jeopardy / bugs here.  I think @chrisr3 was mentioning some.  My main concern is that the `String` of the ID is new every time.  So if someone requests the same attachment in parallel, one potentially overwrites the other and then the lifetimes of the `String` in the `WeakHashMap` are not aligned with the need to keep the mapping in the map and you could find that it's not possible to resolve a URL that should be in the map.

If we could somehow arrange to retrieve the identical `String` that is already a key in the map and use that in the URL then we would be linking the life span appropriately.  But this linking by `String` of the URL path seems very brittle to me.

Basically, it's a right pain.

Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: I've pushed what I think it should look like in a branch `parkri/CORDA-3755`.  It looks a bit mad, but I think this is required.

**Review by rick-r3 — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 5 above)
Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Basically I created a strong reference chain from the `Attachment` to the `String` ID.  The `Attachment` need to be shared always, so there's a de-duplication step there, and the de-duplicated ones are retained by the class loader to assure equal or better life span.  Finally the original map's value is wrapped in a `WeakReference` otherwise there's a cycle.  It might have needed it before, perhaps, but it definitely does now.

Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Two threads (lines begin with thread number):

- 1. Build attachment class loader with attachments (A, B)
- 1. Map ends up with "A" = A and "B" = B
- 2. Build attachment class loader with attachments (A, C) 
- 2. Map ends up with "A"(prime) = A, "B" = B, "C" = C (i.e. the string "A" is different instance)
- 2. Looks up attachments during verify .... all good.
- 2. Class loader discarded (I know it's cached.... but might be evicted and want it to work without cache). "A"(prime) goes out of scope so attachment is discarded from map.
- 1. Looks up attachment for "A", finds null.  Boom!

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 6]
```diff
@@ -359,43 +352,87 @@ object AttachmentsClassLoaderBuilder {
...
+}
+
+interface AttachmentsHolder {
+    val size: Int
+    fun getKey(key: URL): URL?
+    operator fun get(key: URL): Attachment?
+    operator fun set(key: URL, value: Attachment)
+}
+
+private class AttachmentsHolderImpl : AttachmentsHolder {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt:405]: Nice.  Really boosts readability at the call site.

**Review by rick-r3 — DISMISSED** [review]
LGTM

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 7]
```diff
@@ -120,9 +127,128 @@ class ClassLoadingUtilsTest {
         }
     }
 
-    private fun signedAttachment(data: ByteArray, vararg parties: Party) = ContractAttachment.create(
+    // Ignoring as using System.gc in this test which has no guarantees when/if gc occurs. 
+    @Ignore
```
Comment (chrisr3) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: Even better:
```kotlin
@Ignore("Using System.gc in this test which has no guarantees when/if gc occurs.")
```

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 8]
```diff
@@ -120,9 +127,128 @@ class ClassLoadingUtilsTest {
...
+        url1 = null
+        System.gc()
+        val ref1 = referenceQueue1.remove(500)
+        assertNull(ref1)
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+
+        url2 = null
+        System.gc()
+        val ref2 = referenceQueue2.remove(100000)
+        assertNotNull(ref2)
```
Comment (chrisr3) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]: More accurately:
```kotlin
assertSame(weakReference2, ref2)
assertSame(weakReference1, referenceQueue1.poll())
```
And then you can suppress fewer warnings too:
```kotlin
@Suppress("ExplicitGarbageCollectionCall", "UNUSED_VALUE")
```

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 9]
```diff
@@ -120,9 +127,128 @@ class ClassLoadingUtilsTest {
...
+        val weakReference1 = WeakReference<URL>(url1, referenceQueue1)
+
+        val referenceQueue2: ReferenceQueue<URL> = ReferenceQueue()
+        val weakReference2 = WeakReference<URL>(url2, referenceQueue2)
+
+        assertEquals(2, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+        url1 = null
+        System.gc()
+        val ref1 = referenceQueue1.remove(100000)
+        assertNotNull(ref1)
```
Comment (chrisr3) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]:
```kotlin
assertSame(weakReference1, ref1)
```

File: core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt [hunk 10]
```diff
@@ -120,9 +127,128 @@ class ClassLoadingUtilsTest {
...
+        assertEquals(1, AttachmentURLStreamHandlerFactory.loadedAttachmentsSize())
+
+        url2 = null
+        System.gc()
+        val ref2 = referenceQueue2.remove(100000)
+        assertNotNull(ref2)
```
Comment (chrisr3) [inline core/src/test/kotlin/net/corda/core/internal/ClassLoadingUtilsTest.kt]:
```kotlin
assertSame(weakReference2, ref2)
```
And then suppress fewer warnings again.

**Review by chrisr3 — COMMENTED** [review]
Just a few comments on the tests.

**Comment by rick-r3** [comment]
I think if you rebase this you might pick up some test fixes

**Review by rick-r3 — APPROVED** [review]
